### PR TITLE
MIGRATIONS-1080 - Stop directly using an HttpObjectDecoder

### PR DESCRIPTION
**Pull request by gregschohn**
### Description
* **_Category_**: Bug fix
* _**Why these changes are required**_?: To fix bugs when packet boundaries may be anywhere within an HTTP request.
* **_What is the old behavior before changes and new behavior after changes?_**: The TrafficStream recording would be incomplete and/or we would indefinitely hang on requests to ConditionallyReliableLoggingHttpRequestHandler that should have been stopped for the last bytes before flushing.  The old code would get confused and the state model would break.

This change replaces direct usage an HttpObjectDecoder for an HttpRequestDecoder.

ByteBufs are now passed through an EmbeddedChannel that has an HttpRequestDecoder (SImpleHttpRequestDecoder to be specific, which is a bit lighter on memory retention) followed by a SimpleDecodedHttpRequestHandler. When the second handler gets notification that the HttpRequest and LastHttpContent objects are received, that triggers different callbacks. This allows the HttpRequestDecoder object to do the intra-line buffering that the Decoder by itself doesn't do. We're still able to use the decoder in a lightweight way by overriding createMessage() to use a new PassThroughHeaders as opposed to DefaultHttpHeaders(). Both of the tests within ConditionallyReliableLoggingHttpRequestHandlerTest now pass.

### Issues Resolved
MIGRATIONS-1080

### Testing
Unit tests pass.  :dockerSolution:composeUp shows tuples being output by the comparator w/ responses (spot-checked).

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java [hunk 1]
```diff
@@ -19,75 +23,85 @@
...
-         *
-         * @param ctx    the {@link ChannelHandlerContext} which this {@link ByteToMessageDecoder} belongs to
-         * @param buffer the {@link ByteBuf} from which to read data
-         * @param out    the {@link List} to which decoded messages should be added
-         * @throws Exception
-         */
         @Override
-        public void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) throws Exception {
-            super.decode(ctx, buffer, out);
+        public HttpMessage createMessage(String[] initialLine) throws Exception {
```
Comment (lewijacn) [inline TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:27]: nit: I know this is existing code but may be helpful to callout the benefit of using PassThruHttpHeaders here

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java [hunk 2]
```diff
@@ -19,75 +23,85 @@
...
+            currentRequest = null;
+            return old;
         }
     }
 
     protected final IChannelConnectionCaptureSerializer trafficOffloader;
 
-    protected final SimpleHttpRequestDecoder httpDecoder;
+    protected final EmbeddedChannel httpDecoderChannel;
+    SimpleHttpRequestDecoder requestDecoder;
```
Comment (lewijacn) [inline TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:64]: Can this be `protected final` as well ?

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java [hunk 3]
```diff
@@ -19,75 +23,85 @@
...
-        httpDecoder = new SimpleHttpRequestDecoder();
-        currentHttpRequest = null;
+        requestDecoder = new SimpleHttpRequestDecoder(); // as a field for easier debugging
+        httpDecoderChannel = new EmbeddedChannel(
+                requestDecoder,
+                new SimpleDecodedHttpRequestHandler()
+        );
     }
 
+
```
Comment (lewijacn) [inline TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:76]: nit: extra lines here

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java [hunk 4]
```diff
@@ -157,6 +171,7 @@ public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exceptio
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
         trafficOffloader.addExceptionCaughtEvent(Instant.now(), cause);
+        httpDecoderChannel.close();
```
Comment (lewijacn) [inline TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:174]: Do we need to close on normal termination as well?

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java [hunk 5]
```diff
@@ -19,75 +23,85 @@
...
-            httpDecoder.decode(ctx, bb, parsedMsgs);
-            rval = onHttpObjectsDecoded(parsedMsgs);
-            if (rval == HttpCaptureSerializerUtil.HttpProcessedState.FULL_MESSAGE || // got what we needed
-                    lastIdx == bb.readerIndex()) { // no progress
-                break;
-            }
-            parsedMsgs.clear();
-        }
-        log.trace("Resetting index");
+        httpDecoderChannel.writeInbound(bb);
```
Comment (lewijacn) [inline TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:96]: +1 this method seems easier to follow now

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java [hunk 6]
```diff
@@ -19,75 +23,85 @@
...
         bb.resetReaderIndex();
-        return rval;
+        return getHttpRequestReceiverHandler().isDone ?
+                HttpCaptureSerializerUtil.HttpProcessedState.FULL_MESSAGE :
+                HttpCaptureSerializerUtil.HttpProcessedState.ONGOING;
+    }
+
+    private SimpleDecodedHttpRequestHandler getHttpRequestReceiverHandler() {
```
Comment (lewijacn) [inline TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:103]: nit: the naming being different from type and method name threw me off a little

**Review by lewijacn — COMMENTED** [review]
Overall this seems a bit easier to follow now. One of my main open questions is do we need this logic on the HttpResponseHandler as well? Left some minor comments as well

File: TrafficCapture/trafficCaptureProxyServer/src/test/resources/log4j2.properties
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficCaptureProxyServer/src/test/resources/log4j2.properties:1]: Thanks... Unintentional checkin. (the following are notes mostly for myself or anybody else that wants to think about how logging for unit tests should work).

However, in tweaking some code in response to later comments, I went back and forth between whether unit tests should output to stderr and stdout.  I'm not sure that stdout is superior from a developer perspective, especially since our tools (though not this one) output a lot to stdout for machine consumption.  

Either way, gradle won't show the output of stderr or stdout unless you have `test.testLogging.showStandardStreams=true` or run with gradle test with the `-i` option.  In either of those cases BOTH stdout and stderr will render - so it feels like keeping this the same as the main jar is the easiest way to go.  It also lets us test logging configurations.  I've also tested that the tests will pickup the log configuration in the project's jar file if there isn't one in the test jar artifact.

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java (hunk 4 above)
Comment (gregschohn) [reply TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:162]: It shouldn't hurt, provided that we can identify what normal termination is - and I'm not 100% sure yet what exactly those cases are.  This close() is here not for resource management but because the connection and this handler should be in a terminal, non-recoverable state.  If we prematurely close the channel though and then have more data to write, that would be bad.  Given that the EmbeddedChannel doesn't hold any disposable resources, I'm fine letting the GC take the channel, pipeline, & handlers all out when this object gets finalized.  

I plan to address stress testing through many of the corner cases in the coming weeks, but it's hard to write representative tests without seeing some real traffic flow through the system first.

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java (hunk 1 above)
Comment (gregschohn) [reply TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:31]: Done.  Thanks

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java [hunk 7]
```diff
@@ -73,19 +73,7 @@ public LoggingHttpRequestHandler(IChannelConnectionCaptureSerializer trafficOffl
...
-
-
-    protected HttpCaptureSerializerUtil.HttpProcessedState
-    onHttpObjectsDecoded(List<Object> parsedMsgs) throws IOException {
-        parsedMsgs.stream()
-                .filter(o -> o instanceof DefaultHttpRequest)
-                .map(o -> (DefaultHttpRequest) o)
-                .findAny()
-                .ifPresent(req -> getHttpRequestReceiverHandler().currentRequest = req);
-        return HttpCaptureSerializerUtil.addRelevantHttpMessageIndicatorEvents(trafficOffloader, parsedMsgs);
```
Comment (lewijacn) [inline TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:68]: Don't we still need to call `addRelevantHttpMessageIndicatorEvents` ?

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java (hunk 6 above)
Comment (lewijacn) [reply TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java]: Slightly, my inclination is to use `getSimpleDecodedHttpRequestHandler` but that might not be clear to others so I'm fine either way

File: TrafficCapture/trafficCaptureProxyServer/src/test/java/org/opensearch/migrations/trafficcapture/InMemoryConnectionCaptureFactory.java [hunk 8]
```diff
@@ -48,6 +50,7 @@ public IChannelConnectionCaptureSerializer createOffloader(String connectionId)
             CompletableFuture cf = closeHandler(codedStreamToByteBufferMap.get(codedOutputStream));
             codedStreamToByteBufferMap.remove(codedOutputStream);
             singleAggregateCfRef[0] = singleAggregateCfRef[0].isDone() ? cf : CompletableFuture.allOf(singleAggregateCfRef[0], cf);
+            cf.whenComplete((v,t)->onCaptureClosedCallback.run());
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficCaptureProxyServer/src/test/java/org/opensearch/migrations/trafficcapture/InMemoryConnectionCaptureFactory.java:53]: Quick answer, what was specified in the constructor.  In our case for the unit test, it decrements the countdownLatch upon each completion.  Notice that this class is just for test purposes (it was a bit weird to me to chain yet one more listener to bind this to the sending context).

[the remark this answers is not in the document]
Comment (lewijacn) [reply TrafficCapture/trafficCaptureProxyServer/src/test/java/org/opensearch/migrations/trafficcapture/InMemoryConnectionCaptureFactory.java:53]: Okay makes sense now, I didn't pick up this was being passed into the constructor

File: TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java (hunk 7 above)
Comment (gregschohn) [reply TrafficCapture/nettyWireLogging/src/main/java/org/opensearch/migrations/trafficcapture/netty/LoggingHttpRequestHandler.java:68]: Yep - next version does that.  It will include an additional check in a unit test to make sure that the EOM is there for the requests.

**Review by gregschohn — COMMENTED** [review]
To answer the question about "do we need this for responses".  See [MIGRATIONS-1114](https://opensearch.atlassian.net/jira/software/c/projects/MIGRATIONS/boards/1?modal=detail&selectedIssue=MIGRATIONS-1114).  Trying to do any kind of HTTP parsing and EOM tracking for responses is futile anyway.  I'm going to be undertaking 1114 to just put that responsibility in the replayer.

**Review by lewijacn — APPROVED** [review]
Thanks for updates
